File: LeetCode/Liang/109ConvertSortedListToBinarySearchTree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST_Liang(ListNode* head) {
        if (!head) {
            return nullptr;
        }
        
        if (!head->next) {
            return new TreeNode(head->val);
        }
        
        // Find mid point
        ListNode *dummy = new ListNode(INT_MIN);
        dummy->next = head;
        ListNode *preMid = dummy, *end = dummy;
        while (end && end->next) {
            end = end->next->next;
            if (!end) {
                break;
            }
            preMid = preMid->next;
        }
        
        TreeNode *root = new TreeNode(preMid->next->val);
        ListNode *secondHalf = preMid->next->next;
        preMid->next = nullptr;
        ListNode *firstHalf = dummy->next;
        root->left = sortedListToBST(firstHalf);
        root->right = sortedListToBST(secondHalf);
        delete dummy;
        return root;
    }
};

File: LeetCode/Hao/109ConvertSortedListToBinarySearchTree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    ListNode node;
    private TreeNode inorder(int size) {
        if (size == 0) {
            return null;
        }
        
        int leftSize = size / 2;
        int rightSize = size - leftSize - 1;
        
        TreeNode leftTree = inorder(leftSize);
        TreeNode root = new TreeNode(node.val);
        node = node.next;
        TreeNode rightTree = inorder(rightSize);
        
        root.left = leftTree;
        root.right = rightTree;
        
        return root;
    }
    public TreeNode sortedListToBST(ListNode head) {
        int count = 0;
        for (ListNode p = head; p != null ; p = p.next) {
            count ++;
        }
        this.node = head;
        
        return inorder(count);
    }
}
